File: yl-common/src/main/java/com/yoloo/server/common/entity/BaseEntity.java
package com.yoloo.server.common.entity;

import com.googlecode.objectify.annotation.OnLoad;
import com.googlecode.objectify.annotation.OnSave;
import com.yoloo.server.common.vo.AuditData;
import com.yoloo.server.common.vo.Keyable;

import java.time.LocalDateTime;

public abstract class BaseEntity<I, E> implements Entity<I, E>, Keyable<E> {

  private AuditData auditData;

  public BaseEntity() {}

  public AuditData getAuditData() {
    return auditData;
  }

  @OnSave
  protected void onSave() {
    if (auditData == null) {
      auditData = new AuditData(LocalDateTime.now(), null);
    }
  }

  @OnLoad
  protected void onLoad() {}
}

File: services/core-service/src/main/kotlin/com/yoloo/server/entity/BaseEntity.kt
package com.yoloo.server.entity

import com.googlecode.objectify.annotation.OnLoad
import com.googlecode.objectify.annotation.OnSave
import com.yoloo.server.common.vo.AuditData

abstract class BaseEntity<E>(var auditData: AuditData = AuditData()) : Keyable<E> {

    @OnSave
    protected open fun onSave() {
    }

    @OnLoad
    protected open fun onLoad() {
    }
}
